devito/arguments.py

```diff
@@ -1,152 +1,133 @@
+                       to happen.
 """
 
 
-class Argument(object):
+class Parameter(object):
+    """ Abstract base class for any object that represents a node in the dependency
+        graph. It may or may not represent a runtime argument.
 
-    """ Abstract base class for any object that represents a run time argument for
-        generated kernels.
+    :param name: Name of the parameter
+    :param dependencies: A list of :class:`Dependency` objects that represent all the
+                         incoming edges into this node of the graph.
     """
-
-    __metaclass__ = abc.ABCMeta
-
+    is_Argument = False
     is_ScalarArgument = False
     is_TensorArgument = False
     is_PtrArgument = False
 
-    def __init__(self, name, provider, default_value=None):
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, name, dependencies):
         self.name = name
-        self.provider = provider
-        self.default_value = default_value
-        self.reset()
+        self.dependencies = dependencies
 
     @property
-    def value(self):
-        try:
-            if self._value.is_SymbolicFunction:
-                return self._value._data_buffer
-            elif self._value.is_Constant:
-                return self._value.data
-            else:
-                raise InvalidArgument("Unexpected data object %s" % type(self._value))
-        except AttributeError:
-            return self._value
+    def gets_value_from(self):
+        return [x for x in self.dependencies if x.dependency_type == "gets_value_from"]
 
     @property
-    def dtype(self):
-        return self.provider.dtype
+    def verified_by(self):
+        return [x for x in self.dependencies if x.dependency_type == "verified_by"]
 
-    def reset(self):
-        self._value = self.default_value
+    def __repr__(self):
+        return self.name + ", Depends: " + str(self.dependencies)
 
-    @abc.abstractproperty
-    def verify(self, kwargs):
-        return
 
-    def __repr__(self):
-        return self.name
+class DimensionParameter(Parameter):
+    """ Parameter object (node in the dependency graph) that represents a
+        :class:`Dimension`. A dimension object plays an important role in value derivation
+        and verification but does not represent a runtime argument itself (since it
+        provides multiple ScalarArguments).
+    """
+    def __init__(self, provider, dependencies):
+        super(DimensionParameter, self).__init__(provider.name, dependencies)
+        self.provider = provider
+
+
+class Argument(Parameter):
+
+    """ Base class for any object that represents a run time argument for
+        generated kernels. It is necessarily a node in the dependency graph.
+    """
+    is_Argument = True
+
+    def __init__(self, name, dependencies, dtype=np.int32):
+        super(Argument, self).__init__(name, dependencies)
+        self.dtype = dtype
 
 
 class ScalarArgument(Argument):
```

Comment:
So these still don't have providers?

Reply:
No. Does not make sense to me for `ScalarArgument`s to have providers.

Reply:
why?

Reply:
TensorArguments come from objects that are symbols but have data attached to them - that's their provider. ScalarArguments don't have any such "provider". Perhaps you could explain why you think they do.

Reply:
In the current design, the `ValueVisitor` is deriving the values so a provider isn't necessary. We can discuss it for the future possibly.